masonry/src/render_root.rs

```diff
@@ -295,83 +306,94 @@ impl RenderRoot {
+            // Our WidgetArena stores all widgets as Box<dyn Widget>, but the "true"
+            // type of our root widget is *also* Box<dyn Widget>. We downcast so we
+            // don't add one more level of indirection to this.
+            let widget = widget_mut
+                .widget
+                .as_mut_dyn_any()
+                .downcast_mut::<Box<dyn Widget>>()
+                .unwrap();
+            let widget_mut = WidgetMut {
+                ctx: widget_mut.ctx.reborrow_mut(),
+                widget,
+            };
+            f(widget_mut)
+        });
+
+        let mut root_state = self.widget_arena.get_state_mut(self.root.id()).item.clone();
+        self.post_event_processing(&mut root_state);
+
+        res
+    }
 
+    /// Get a [`WidgetMut`] to a specific widget.
+    ///
+    /// Because of how `WidgetMut` works, it can only be passed to a user-provided callback.
+    pub fn edit_widget<R>(
+        &mut self,
+        id: WidgetId,
+        f: impl FnOnce(WidgetMut<'_, Box<dyn Widget>>) -> R,
+    ) -> R {
+        // TODO - Factor out into a "pre-event" function?
         self.state.next_focused_widget = self.state.focused_widget;
-        let root_widget = WidgetMut {
-            ctx: WidgetCtx {
-                global_state: &mut self.state,
-                parent_widget_state: &mut fake_widget_state,
-                widget_state: state_ref.item,
-                widget_state_children: state_ref.children,
-                widget_children: widget_ref.children,
-            },
-            widget,
-            is_reborrow: false,
-        };
 
-        let res = {
-            let _span = info_span!("edit_root_widget").entered();
-            f(root_widget)
-        };
-        self.post_event_processing(&mut fake_widget_state);
+        let mut res = None;
+        mutate_widget(self, id, |widget_mut| {
+            res = Some(f(widget_mut));
+        });
 
-        res
+        let mut root_state = self.widget_arena.get_state_mut(self.root.id()).item.clone();
+        self.post_event_processing(&mut root_state);
+
+        res.unwrap()
     }
 
     // --- MARK: POINTER_EVENT ---
     fn root_on_pointer_event(&mut self, event: PointerEvent) -> Handled {
-        let mut root_state =
-            WidgetState::new(self.root.id(), Some(self.get_kurbo_size()), "<root>");
+        let mut dummy_state = WidgetState::synthetic(self.root.id(), self.get_kurbo_size());
 
-        let handled = root_on_pointer_event(self, &mut root_state, &event);
-        run_update_pointer_pass(self, &mut root_state);
+        // TODO - Factor out into a "pre-event" function?
+        self.state.next_focused_widget = self.state.focused_widget;
 
-        self.post_event_processing(&mut root_state);
+        let handled = root_on_pointer_event(self, &mut dummy_state, &event);
+        run_update_pointer_pass(self, &mut dummy_state);
+
+        self.post_event_processing(&mut dummy_state);
```

Comment:
It's not clear to me what role we expect `post_event_processing` to have "long-term".

Reply:
I'll remove it once all the passes are split out from RenderRoot and WidgetPod.